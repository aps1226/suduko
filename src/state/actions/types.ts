export const SET_BOARD: string = 'SET_BOARD';

export const CHANGE_COLOR: string = 'CHANGE_COLOR';

export const SET_SELECTION: string = 'SET_SELECTION';