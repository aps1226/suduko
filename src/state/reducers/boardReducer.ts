import { AnyAction } from 'redux';
import { SET_BOARD } from '../actions/types';
import { shuffle } from '../boardController';

const initBoard:(number | null)[][] = [shuffle([1,2,3,4,5,6,7,8,9]),[null,null,null,null,null,null,null,null,null],[null,null,null,null,null,null,null,null,null],[null,null,null,null,null,null,null,null,null],[null,null,null,null,null,null,null,null,null],[null,null,null,null,null,null,null,null,null],[null,null,null,null,null,null,null,null,null],[null,null,null,null,null,null,null,null,null],[null,null,null,null,null,null,null,null,null]];

const boardReducer = (state = initBoard, action: AnyAction):(number | null)[][]  =>{
  switch(action.type){
    case SET_BOARD:
      return action.payload;
    default:
      return state;
  }
}

export default boardReducer;