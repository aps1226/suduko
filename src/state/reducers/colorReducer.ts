import {CHANGE_COLOR,DEFAULT_COLORS} from '../actions/types';

import {Colors} from '../../../types'

const initGradient = ():string[] =>{
  const woodPalate:string[] = ["#a37e5c", "#b89372","#c8a484","#dcb999", "#d3af8f","#d9b696","#ceaa89","#d2ae8d","#b38e6d","aa8563"];
  const gradient:string[] = [];
  for(let i = 0; i < 100; i++){
    gradient.push(woodPalate[Math.round(Math.random()*10)])
  }
  return gradient;
}

const initializedGradient:string[] = initGradient();

const initialState:Colors = {
  1:['transparent','black'],
  2:['transparent','black'],
  3:['transparent','black'],
  4:['transparent','black'],
  5:['transparent','black'],
  6:['transparent','black'],
  7:['transparent','black'],
  8:['transparent','black'],
  9:['transparent','black'],
  'Selection_1':'transparent',
  'Selection_2':'transparent',
  'Selection_3':'transparent',
  'Selection_4':'transparent',
  'Selection_5':'transparent',
  'Selection_6':'transparent',
  'Selection_7':'transparent',
  'Selection_8':'transparent',
  'Selection_9':'transparent',
  'gradient': initializedGradient
}
const colorReducer = (state = initialState, action:any) =>{

  const initialState:Colors = {
    1:['transparent','black'],
    2:['transparent','black'],
    3:['transparent','black'],
    4:['transparent','black'],
    5:['transparent','black'],
    6:['transparent','black'],
    7:['transparent','black'],
    8:['transparent','black'],
    9:['transparent','black'],
    'Selection_1':'transparent',
    'Selection_2':'transparent',
    'Selection_3':'transparent',
    'Selection_4':'transparent',
    'Selection_5':'transparent',
    'Selection_6':'transparent',
    'Selection_7':'transparent',
    'Selection_8':'transparent',
    'Selection_9':'transparent',
    'gradient': initializedGradient
  }
  
  switch(action.type){
    case CHANGE_COLOR:
      if(!action.payload.includes('Selection')) initialState[action.payload] = ['#008000','#008000'];
      else initialState[action.payload] = '#3297FD';
      return initialState;
    case DEFAULT_COLORS:
      return initialState;
    default:
      return state;
  }
}

export default colorReducer;