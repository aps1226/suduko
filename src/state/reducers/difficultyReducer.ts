import { SET_DIFFICULTY } from '../actions/types';
const difficultyReducer = (state = null, action: any) =>{
  switch(action.type){
    case SET_DIFFICULTY:
      return action.payload;
    default:
      return state;
  }
}

export default difficultyReducer;