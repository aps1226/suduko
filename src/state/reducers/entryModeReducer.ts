import {SET_ENTRY_MODE} from '../actions/types';

const entryModeReducer = (state = true, action: any) =>{
  switch(action.type){
    case SET_ENTRY_MODE:
      return action.payload;
    default:
      return state;
  }
}

export default entryModeReducer;