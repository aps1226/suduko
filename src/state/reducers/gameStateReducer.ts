import { AnyAction } from 'redux';
import { SET_GAME_STATE } from '../actions/types';
import { GameState } from '../../../types';

const initGameState:GameState = {
  'isCompleted':false,
  'gameExists':false
}
const gameStateReducer = (state = initGameState, action: AnyAction):GameState =>{
  switch(action.type){
    case SET_GAME_STATE:
      return action.payload;
    default:
      return state;
  }
}

export default gameStateReducer;