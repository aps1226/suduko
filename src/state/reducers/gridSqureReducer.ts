import {INCREMENT} from '../actions/types';
const gridSquareReducer = (state = [0,0], action: any) =>{
  switch(action.type){
    case INCREMENT:
      const curPosition:number[] = state;
      console.log('increment called',curPosition)
      if(curPosition[1] < 8) curPosition[1] += 1;
      else {
        curPosition[0] += 1;
        curPosition[1] = 0;
      }
      return curPosition;
    default:
      return state;
  }
}

export default gridSquareReducer;