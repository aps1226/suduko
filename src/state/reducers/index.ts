import {combineReducers} from 'redux';
import boardReducer from './boardReducer';
import colorReducer from './colorReducer';
import selectionReducer from './selectionReducer';

const reducers = combineReducers({
  board: boardReducer,
  colors: colorReducer,
  selection:selectionReducer
});

export default reducers;