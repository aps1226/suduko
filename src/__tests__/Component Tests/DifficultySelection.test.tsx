import React from 'react';
import renderWithRedux from '../renderWithRedux';
import { createMemoryHistory, createLocation, MemoryHistory, Location} from 'history';
import { match } from 'react-router';
import { act, cleanup, fireEvent, render, waitFor} from '@testing-library/react-native';
import { RenderReduxAPI, MockStore } from '../../../types';
import {RootState} from '../../state/reducers/index';
import DifficultySelection from '../../../components/DifficultySelection';

//Mock props for routing.
const history = createMemoryHistory();
const path:string = `/route/:id`;
const match: match<{ id: string }> = {
    isExact: false,
    path,
    url: path.replace(':id', '1'),
    params: { id: '1' }
};
const location = createLocation(match.url);

//DifficultySelection component test.
describe('DifficultySelection',() =>{
  //Mock store object for state management.
  const mockStore:MockStore = {};
  //Render DifficultySelection component with routing and redux.
  let difficultySelection:RenderReduxAPI;
  beforeEach(async() =>{
    await waitFor(() =>{
      difficultySelection = renderWithRedux(
        <DifficultySelection
          history={history}
          location={location}
          match={match}
        />,
        {
          initialState:mockStore
        }
      );
    })
  })

  afterEach(cleanup);

  //Component renders without crashing.
  it('renders without crashing',()=>{
    expect(difficultySelection.toJSON()).toMatchSnapshot();
  })

  //Test easy selection.
  it(`correctly initializes board when the easy button is selected`, async ()=>{
    const {getByTestId, store} = difficultySelection;
    const easyButton = getByTestId('easyButton');
    const curHistory:number = history.length;
    //Press easy button
    fireEvent.press(easyButton);
    await waitFor(()=>{
      //History length should increase by 1.
      expect(history.length).toBe(curHistory + 1);
      //Location pathname should be 'GameDisplay'.
      expect(history.location.pathname).toBe('/GameDisplay');
      //Determine the amount of non-null values on the initialized board.
      const boardState:(number | null)[][] = store.getState().board;
      let valueCount:number = 0;
      boardState.forEach((arr) =>{
        arr.forEach((num)=>{
          if(num !== null) valueCount += 1;
        })
      })
      //Board should be initialized with (37,46) non-null values.
      expect(valueCount >= 37 && valueCount <= 46).toBeTruthy();
    })
  })

  //Test medium selection.
  it(`correctly initializes board when the medium button is selected`, async ()=>{
    const {getByTestId, store} = difficultySelection;
    const mediumButton = getByTestId('mediumButton');
    const curHistory:number = history.length;
    //Press medium button.
    fireEvent.press(mediumButton);
    await waitFor(()=>{
      //History length should increase by 1.
      expect(history.length).toBe(curHistory + 1);
      //Location pathname should be 'GameDisplay'.
      expect(history.location.pathname).toBe('/GameDisplay');
      //Determine the amount of non-null values on the initialized board.
      const boardState:(number | null)[][] = store.getState().board;
      let valueCount:number = 0;
      boardState.forEach((arr) =>{
        arr.forEach((num)=>{
          if(num !== null) valueCount += 1;
        })
      })
      //Board should be initialized with (27,36) non-null values.
      expect(valueCount >= 27 && valueCount <= 36).toBeTruthy();
    })
  })

  //Test hard selection.
  it(`correctly initializes board when the hard button is selected`, async ()=>{
    const {getByTestId, store} = difficultySelection;
    const hardButton = getByTestId('hardButton');
    const curHistory = history.length;
    //Press hard button
    fireEvent.press(hardButton);
    await waitFor(()=>{
      //History length should increase by 1.
      expect(history.length).toBe(curHistory + 1);
      //Location pathname should be 'GameDisplay'.
      expect(history.location.pathname).toBe('/GameDisplay');
      //Determine the amount of non-null values on the initialized board.
      const boardState:(number | null)[][] = store.getState().board;
      let valueCount:number = 0;
      boardState.forEach((arr) =>{
        arr.forEach((num)=>{
          if(num !== null) valueCount += 1;
        })
      })
      //Board should be initialized with (19,26) non-null values.
      expect(valueCount >= 19 && valueCount <= 26).toBeTruthy();
    })
  })

  //Test extreme selection.
  it(`correctly initializes board when the extreme button is selected`, async ()=>{
    const {getByTestId, store} = difficultySelection;
    const extremeButton = getByTestId('extremeButton');
    const curHistory:number = history.length;
    //Press hard button
    fireEvent.press(extremeButton);
    await waitFor(()=>{
      //History length should increase by 1.
      expect(history.length).toBe(curHistory + 1);
      //Location pathname should be 'GameDisplay'.
      expect(history.location.pathname).toBe('/GameDisplay');
      //Determine the amount of non-null values on the initialized board.
      const boardState:(number | null)[][] = store.getState().board;
      let valueCount:number = 0;
      boardState.forEach((arr) =>{
        arr.forEach((num)=>{
          if(num !== null) valueCount += 1;
        })
      })
      //Board should be initialized with 18 non-null values.
      expect(valueCount).toBe(18);
    })
  })

  //Test main menu button.
  it(`should route to the home screen when the 'Main Menu' button is selected`, ()=>{
    const {getByTestId} = difficultySelection;
    const mainMenuButton = getByTestId('mainMenuButton');
    const curHistory = history.length;
    //Press 'Main Menu' button.
    fireEvent.press(mainMenuButton);
    //History length should increase by 1.
    expect(history.length).toBe(curHistory + 1);
    //Location pathname should be '/'.
    expect(history.location.pathname).toBe('/');
  })

})