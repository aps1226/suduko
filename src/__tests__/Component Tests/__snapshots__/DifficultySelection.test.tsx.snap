// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DifficultySelection renders without crashing 1`] = `
<View
  style={
    Object {
      "alignItems": "center",
      "flex": 1,
      "justifyContent": "center",
    }
  }
>
  <View
    style={
      Object {
        "flexDirection": "row",
        "justifyContent": "space-evenly",
        "left": 0,
        "position": "absolute",
        "top": 50,
      }
    }
  >
    <View
      accessible={true}
      focusable={true}
      onBlur={[Function]}
      onClick={[Function]}
      onFocus={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "alignItems": "center",
          "backgroundColor": "#1E90FF",
          "borderRadius": 4,
          "elevation": 3,
          "justifyContent": "center",
          "marginTop": 30,
          "paddingVertical": 12,
          "shadowColor": "rgba(0,0,0,0.35)",
          "shadowOffset": Object {
            "height": -2,
            "width": -2,
          },
          "shadowOpacity": 0.75,
          "shadowRadius": 2,
          "width": 125,
        }
      }
      testID="mainMenuButton"
    >
      <Text
        style={
          Object {
            "color": "white",
            "fontSize": 16,
            "fontWeight": "bold",
            "letterSpacing": 0.25,
            "lineHeight": 21,
          }
        }
      >
        Main Menu
      </Text>
    </View>
  </View>
  <View
    accessible={true}
    focusable={true}
    onBlur={[Function]}
    onClick={[Function]}
    onFocus={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "alignItems": "center",
        "backgroundColor": "#1E90FF",
        "borderRadius": 4,
        "elevation": 3,
        "justifyContent": "center",
        "marginTop": 30,
        "paddingVertical": 12,
        "shadowColor": "rgba(0,0,0,0.35)",
        "shadowOffset": Object {
          "height": -2,
          "width": -2,
        },
        "shadowOpacity": 0.75,
        "shadowRadius": 2,
        "width": 150,
      }
    }
    testID="easyButton"
  >
    <Text
      style={
        Object {
          "color": "white",
          "fontSize": 16,
          "fontWeight": "bold",
          "letterSpacing": 0.25,
          "lineHeight": 21,
        }
      }
    >
      Easy
    </Text>
  </View>
  <View
    accessible={true}
    focusable={true}
    onBlur={[Function]}
    onClick={[Function]}
    onFocus={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "alignItems": "center",
        "backgroundColor": "#1E90FF",
        "borderRadius": 4,
        "elevation": 3,
        "justifyContent": "center",
        "marginTop": 30,
        "paddingVertical": 12,
        "shadowColor": "rgba(0,0,0,0.35)",
        "shadowOffset": Object {
          "height": -2,
          "width": -2,
        },
        "shadowOpacity": 0.75,
        "shadowRadius": 2,
        "width": 150,
      }
    }
    testID="mediumButton"
  >
    <Text
      style={
        Object {
          "color": "white",
          "fontSize": 16,
          "fontWeight": "bold",
          "letterSpacing": 0.25,
          "lineHeight": 21,
        }
      }
    >
      Medium
    </Text>
  </View>
  <View
    accessible={true}
    focusable={true}
    onBlur={[Function]}
    onClick={[Function]}
    onFocus={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "alignItems": "center",
        "backgroundColor": "#1E90FF",
        "borderRadius": 4,
        "elevation": 3,
        "justifyContent": "center",
        "marginTop": 30,
        "paddingVertical": 12,
        "shadowColor": "rgba(0,0,0,0.35)",
        "shadowOffset": Object {
          "height": -2,
          "width": -2,
        },
        "shadowOpacity": 0.75,
        "shadowRadius": 2,
        "width": 150,
      }
    }
    testID="hardButton"
  >
    <Text
      style={
        Object {
          "color": "white",
          "fontSize": 16,
          "fontWeight": "bold",
          "letterSpacing": 0.25,
          "lineHeight": 21,
        }
      }
    >
      Hard
    </Text>
  </View>
  <View
    accessible={true}
    focusable={true}
    onBlur={[Function]}
    onClick={[Function]}
    onFocus={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "alignItems": "center",
        "backgroundColor": "#1E90FF",
        "borderRadius": 4,
        "elevation": 3,
        "justifyContent": "center",
        "marginTop": 30,
        "paddingVertical": 12,
        "shadowColor": "rgba(0,0,0,0.35)",
        "shadowOffset": Object {
          "height": -2,
          "width": -2,
        },
        "shadowOpacity": 0.75,
        "shadowRadius": 2,
        "width": 150,
      }
    }
    testID="extremeButton"
  >
    <Text
      style={
        Object {
          "color": "white",
          "fontSize": 16,
          "fontWeight": "bold",
          "letterSpacing": 0.25,
          "lineHeight": 21,
        }
      }
    >
      Extreme
    </Text>
  </View>
</View>
`;
